src/doc/trpl/choosing-guarantees.md

```diff
@@ -0,0 +1,417 @@
+
+`Rc<T>` has the computational cost of incrementing/decrementing the refcount whenever it is cloned
+or goes out of scope respectively. Note that a clone will not do a deep copy, rather it will simply
+increment the inner reference count and return a copy of the `Rc<T>`
+
+
+
+
+[rc]: ../std/rc/struct.Rc.html
+
+# Cell types
+
+"Cells" provide interior mutability. In other words, they contain data which can be manipulated even
+if the type cannot be obtained in a mutable form (for example, when it is behind an `&`-ptr or
+`Rc<T>`).
+
+[The documentation for the `cell` module has a pretty good explanation for these][cell-mod].
+
+These types are _generally_ found in struct fields, but they may be found elsewhere too.
+
+## `Cell<T>`
+
+[`Cell<T>`][cell] is a type that provides zero-cost interior mutability, but only for `Copy` types.
+Since the compiler knows that all the data owned by the contained value is on the stack, there's
+no worry of leaking any data behind references (or worse!) by simply replacing the data.
+
+It is still possible to violate your own invariants using this wrapper, so be careful when using it.
+If a field is wrapped in `Cell`, it's a nice indicator that the chunk of data is mutable and may not
+stay the same between the time you first read it and when you intend to use it.
+
+
+
+```rust
+let x = Cell::new(1);
+let y = &x;
+let z = &x;
+x.set(2);
+y.set(3);
+z.set(4);
+println!("{}", x.get());
+```
+
+Note that here we were able to mutate the same value from various immutable references.
+
+
+This has the same runtime cost as the following:
+
+
+```rust
+let mut x = 1;
+let y = &mut x;
+let z = &mut x;
+x = 2;
+*y = 3;
+*z = 4;
+println!("{}", x;
+```
+
+but it has the added benefit of actually compiling successfully.
+
+
+#### Guarantees
+
+This relaxes the "no aliasing with mutability" restriction in places where it's unnecessary.
+However, this also relaxes the guarantees that the restriction provides; so if one's invariants
+depend on data stored within `Cell`, one should be careful.
+
+
+This is useful for mutating primitives and other `Copy` types when there is no easy way of
+doing it in line with the static rules of `&` and `&mut`.
+
+[Gábor Lehel][glaebhoerl] summed up the guarantees provided by `Cell` in a rather succinct manner:
+
+> The basic guarantee we need to ensure is that interior references can't be invalidated (left
+> dangling) by mutation of the outer structure. (Think about references to the interiors of types
+> like `Option`, `Box`, `Vec`, etc.) `&`, `&mut`, and Cell each make a different tradeoff here. `&`
+> allows shared interior references but forbids mutation; `&mut` allows mutation xor interior
+> references but not sharing; `Cell` allows shared mutability but not interior references.
+
+[because `Cell` operates by copying, it can't create interior references]
+
+[glaebhoerl]: http://www.reddit.com/r/rust/comments/378tj6/blog_post_review_requested_container_types/crlfl74
```

Comment:
we generally don't link to anything external in the book, and I _certainly_ dont want a link to reddit